File: platforms/hacker-rank/M--40--absolute-permutation/solutions/solution.py
#!/bin/python3

import math
import os
import random
import re
import sys

#
# Complete the 'absolutePermutation' function below.
#
# The function is expected to return an INTEGER_ARRAY.
# The function accepts following parameters:
#  1. INTEGER n
#  2. INTEGER k
#


def absolutePermutation(n, k):
    if k == 0:
        return list(range(1, n + 1))

    if n % (2 * k) != 0:
        return [-1]

    res = []
    add = True
    for i in range(1, n + 1, k):
        if add:
            for j in range(i, i + k):
                res.append(j + k)
        else:
            for j in range(i, i + k):
                res.append(j - k)
        add = not add

    return res


if __name__ == "__main__":
    fptr = open(os.environ["OUTPUT_PATH"], "w")
    t = int(input().strip())

    for t_itr in range(t):
        first_multiple_input = input().rstrip().split()
        n = int(first_multiple_input[0])
        k = int(first_multiple_input[1])
        result = absolutePermutation(n, k)
        fptr.write(" ".join(map(str, result)))
        fptr.write("\n")

    fptr.close()

File: platforms/hacker-rank/M--40--absolute-permutation/solutions/solution.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"strconv"
	"strings"
)

/*
 * Complete the 'absolutePermutation' function below.
 *
 * The function is expected to return an INTEGER_ARRAY.
 * The function accepts following parameters:
 *  1. INTEGER n
 *  2. INTEGER k
 */

func absolutePermutation(n int32, k int32) []int32 {
	if k == 0 {
		res := make([]int32, n)
		for i := int32(0); i < n; i++ {
			res[i] = i + 1
		}
		return res
	}

	if n%(2*k) != 0 {
		return []int32{-1}
	}

	res := make([]int32, 0, n)
	add := true
	for i := int32(1); i <= n; i += k {
		if add {
			for j := i; j < i+k; j++ {
				res = append(res, j+k)
			}
		} else {
			for j := i; j < i+k; j++ {
				res = append(res, j-k)
			}
		}
		add = !add
	}

	return res
}

func main() {
	reader := bufio.NewReaderSize(os.Stdin, 16*1024*1024)

	stdout, err := os.Create(os.Getenv("OUTPUT_PATH"))
	checkError(err)

	defer stdout.Close()

	writer := bufio.NewWriterSize(stdout, 16*1024*1024)

	tTemp, err := strconv.ParseInt(strings.TrimSpace(readLine(reader)), 10, 64)
	checkError(err)
	t := int32(tTemp)

	for tItr := 0; tItr < int(t); tItr++ {
		firstMultipleInput := strings.Split(strings.TrimSpace(readLine(reader)), " ")

		nTemp, err := strconv.ParseInt(firstMultipleInput[0], 10, 64)
		checkError(err)
		n := int32(nTemp)

		kTemp, err := strconv.ParseInt(firstMultipleInput[1], 10, 64)
		checkError(err)
		k := int32(kTemp)

		result := absolutePermutation(n, k)

		for i, resultItem := range result {
			fmt.Fprintf(writer, "%d", resultItem)

			if i != len(result)-1 {
				fmt.Fprintf(writer, " ")
			}
		}

		fmt.Fprintf(writer, "\n")
	}

	writer.Flush()
}

func readLine(reader *bufio.Reader) string {
	str, _, err := reader.ReadLine()
	if err == io.EOF {
		return ""
	}

	return strings.TrimRight(string(str), "\r\n")
}

func checkError(err error) {
	if err != nil {
		panic(err)
	}
}
